### ISSUE TITLE
Smart Buttons Compatibility with Mix and Match Products and Name Your Price

### ISSUE THREAD

**Issue by helgatheviking** [woocommerce/woocommerce-gateway-paypal-express-checkout#694]
### Describe the bug

Follow up to #554 where I stated that the Smart buttons did work with Name Your Price and Mix and Match. Upon retesting, they both work when you enter a valid configuration/amount in that the correct data is passed to the cart/order. But if the form is not valid the smart buttons are still clickable. You can go to PayPal, login, approve the purchase, and then get directed back to the cart where there's an error and you cannot complete checkout.

I found while digging through #507 that you can trigger the buttons' on/off:

```
$( '#woo_pp_ec_button_product' ).trigger( 'disable' );
$( '#woo_pp_ec_button_product' ).trigger( 'enable' );
```

So that seemed like something I could add on my side when my forms were valid/invalid

However, [this code](https://github.com/woocommerce/woocommerce-gateway-paypal-express-checkout/blob/master/assets/js/wc-gateway-ppec-generate-cart.js#L65-L70)

```
// Disable the button if there are invalid fields in the product page (like required fields from Product Addons)
	form.on( 'change', 'select, input, textarea', function() {
		// Hack: IE11 uses the previous field value for the checkValidity() check if it's called in the onChange handler
		setTimeout( validate_form, 0 );
	} );
	validate_form();
```

Is running your custom validation every time and resetting my changes. As far as I can tell the `checkValidity()` function is only checking if any required inputs are missing... which isn't sufficient for either of my plugins. 

I'm going to dig around a bit more to see if there's something I can do on my side to work within the framework you have, but it might also be helpful if there was an attribute/class that we could toggle to to set the status as enabled/disabled for working with other product types that use different form inputs to configure the product.

### To reproduce

1. Enabled smart buttons on product pages
1. Create a mix and match product
1. do not configure the product, but click on the PayPal pay now button

**Screenshots**

Buttons are enabled when add to cart button is disabled
https://share.getcloudapp.com/geu2Adw4

Return from PayPal but hit order error (since product is not configured correctly)
https://share.getcloudapp.com/4gumlKXW

Buttons are also enabled for Name Your Price and you can complete an order with an amount less than the required minimum. (Min is $49)
https://share.getcloudapp.com/geu2AE9j

At the very least I'll start re-validating NYP amounts on `woocommerce_check_cart_items` which should get the same error result as the MNM screenshot above (go to PayPal and come back to error). 

### Expected behavior

A way to enabled/disable the buttons for custom forms

**Environment (please complete the following information):**
 - WordPress 5.3.02 - WooCommerce 4.0.0-beta
 - PayPal (Express) Checkout Plugin Version 1.6.20
 - FireFox 73.0.1

**Comment by helgatheviking** [comment]
I almost was able to get this resolved by adding custom validity states to my inputs, but my script runs after yours so you'd check the validity of the form, then I'd set the state and it would be too late. 

I think you'll need a little bit of an API so we can interact with the script and override the `fields_valid` variable.

**Comment by james-allan** [comment]
Hi @helgatheviking. Thanks for the brain cycles on this. Would something like this work for you? 

We would trigger a custom JS event to allow you to "filter" the form validity. The changes we could make are: 

```js
// This would be the change made to our code.
var product_form_validity_check = jQuery.Event( 'wc_ppec_valid_product_form' );
event.valid = form.get( 0 ).checkValidity(); // This is the current and default check.

// We trigger the event to allow third parties to filter the event.valid variable.
$( document ).trigger( event );
fields_valid = event.valid;
```

You would then add something like this in your plugin. You would need check the validity of the form and return true or false.
```js
$( document ).on( 'wc_ppec_validate_product_form', function( product_form_validity_check ) {
    if ( true === product_form_validity_check.valid && ! your_validation_condition ) {
        product_form_validity_check.valid = false;
    }
} );


```

**Comment by helgatheviking** [comment]
I like it! I can work with that. Hate having to tell people to disable the smart buttons.

I wonder if I could use the same pattern in one page checkout where I'm having a similar issue.

**Comment by helgatheviking** [comment]
Followup, I have this working about as well as I think I'm gonna get it. there's a small problem when typing the quantity values in. 

https://share.getcloudapp.com/X6uzb6yN

1. you don't listen for `input` and 2. I update the entire container with a little timeout and that's making the status incorrect when the values are entered by typing.

While the PayPal buttons are grayed out they can be clicked on and that triggers the `change` handler and so the buttons are then activated. 

Example page: https://wp-helgatheviking.codeanyapp.com/product/t-shirt-6-pack/

**Comment by helgatheviking** [comment]
@mattallan any idea when this might get shipped?

**Comment by mattallan** [comment]
@helgatheviking This will be shipped with 1.7.0 which is planned to go out sometime during the week beginning 6th APR :)
